### ISSUE TITLE
Cortex failing to handle samples out of nowhere

### ISSUE THREAD

**Issue by jakubgs** [cortexproject/cortex#3513]
## Description

I don't understand what's happening with my Cortex cluster. After several hours of ingesting samples and working fine it just starts failing to receive metrics. One host starts returning `5xx` and `4xx` errors for the push API route and using a lot of CPU.

This issue goes away once I restart the Prometheus instances.

## Setup

* Prometheus `2.22.2` (docker) - 4x `4 CPU 8 GB RAM`
* Cortex `1.5.0` (binary) - 3x `6 CPU 16 GB RAM`
* Cassandra `3.11.9` (binary) - 3x `6 CPU 16 GB RAM`

Here is my Cortex config from one of my hosts: [config](https://gist.github.com/jakubgs/c39915f914be94c7e408083ab2b872cd).

The Cassandra hosts are beefy and quite healthy. Pretty sure that's not the issue.

## Details

If we look at the samples scraped by Prometheus we can see that the number is mostly stable `~13600` samples total:

![prometheus_appended_samples_per_sec](https://user-images.githubusercontent.com/2212681/99661725-8b792280-2a64-11eb-9943-9ea9227cf549.png)

And this matches with number of samples received by each of the Cortex nodes for a while. But then one of the nodes - in this case `master-03` - starts slowly dropping in number of samples, and then has a sharp rise to ~85 K:

![cortex_samples_total_10m_rate](https://user-images.githubusercontent.com/2212681/99662134-24a83900-2a65-11eb-8c1c-d9eb298f31f1.png)

Labels per sample also raise:

![cortex_labels_per_sample_raise](https://user-images.githubusercontent.com/2212681/99662870-1eff2300-2a66-11eb-9e98-45037c32bc26.png)

At the same time we can see that Cortex Distributor is starting to return more and more `5xx` errors on `master-03`:

![cortex_request_responses_5xx_raise](https://user-images.githubusercontent.com/2212681/99662410-7c46a480-2a65-11eb-9cfa-f7b6444dadb2.png)

And this correlates with the `master-03` CPU usage being way above the other 2 nodes in Cortex cluster:

![cortex_hosts_cpu_usage_raise](https://user-images.githubusercontent.com/2212681/99663116-69809f80-2a66-11eb-9f9c-59f33ae0993f.png)

And we can see raise in failures in storing samples on Prometheus for that remote:

![prometheus_samples_issues](https://user-images.githubusercontent.com/2212681/99662698-e4958600-2a65-11eb-8e11-2469eee8bfc5.png)

As well as jump in remote storage shards:

![prometheus_storage_shards](https://user-images.githubusercontent.com/2212681/99662763-faa34680-2a65-11eb-907e-63d871bfe555.png)

## Questions

What is causing this? I did not change anything. It was working fine for a few hours and then suddenly issues started.

* Is this Prometheus sending samples incorrectly?
* Is it Cortex failing to handle the samples?
* Why is just one host in the cluster being overloaded? Shouldn't the load be spread?
* What can I do to diagnose this further?

**Comment by jakubgs** [comment]
I might add that my Prometheus configuration is stock. I do not configure any `queue_config` settings. I just set:
```yaml
remote_write:
  - url: "http://10.1.31.155:9092/api/v1/push" # master-01.do-ams3.metrics.hq
  - url: "http://10.1.32.207:9092/api/v1/push" # master-02.do-ams3.metrics.hq
  - url: "http://10.1.33.0:9092/api/v1/push" # master-03.do-ams3.metrics.hq
```

**Comment by jakubgs** [comment]
Just to be clear, Cassandra is fine. Here's the perfectly healthy write latency for the Cassandra cluster:

![cassandra_write_latency_fine](https://user-images.githubusercontent.com/2212681/99665313-c29e0280-2a69-11eb-9c52-3fc6fa5c8a9a.png)

As is the commit log:

![cassandra_commit_log_fine](https://user-images.githubusercontent.com/2212681/99665483-f9741880-2a69-11eb-9e92-c136c7af9c42.png)

So I'm pretty darn sure the Cassandra cluster is not the issue.

**Comment by jakubgs** [comment]
This metric makes me wonder:

![prometheus_shards_desired](https://user-images.githubusercontent.com/2212681/99665903-8a4af400-2a6a-11eb-99f9-2cbff700fb0f.png)

The metric `prometheus_remote_storage_shards_desired` gradually rises until I restart the Prometheus instances and goes back to healthy `1`(I assume healthy...).

Not sure what to make of this.

**Comment by jakubgs** [comment]
The [Prometheus docs](https://prometheus.io/docs/practices/remote_write/) state that:

>Each remote write destination starts a queue which reads from the write-ahead log (WAL), writes the samples into an in memory queue owned by a shard, which then sends a request to the configured endpoint.
>
>When one shard backs up and fills its queue, Prometheus will block reading from the WAL into any shards. Failures will be retried without loss of data __unless the remote endpoint remains down for more than 2 hours__. After 2 hours, the WAL will be compacted and data that has not been sent will be lost.

The 2 hours might match what I'm seeing. But the thing is Cortex is not down, it works just fine until it doesn't.

**Comment by jakubgs** [comment]
Here are the defaults based on Prometheus docs:
```yaml
# Configures the queue used to write to remote storage.
queue_config:
  # Number of samples to buffer per shard before we block reading of more
  # samples from the WAL. It is recommended to have enough capacity in each
  # shard to buffer several requests to keep throughput up while processing
  # occasional slow remote requests.
  [ capacity: <int> | default = 500 ]
  # Maximum number of shards, i.e. amount of concurrency.
  [ max_shards: <int> | default = 1000 ]
  # Minimum number of shards, i.e. amount of concurrency.
  [ min_shards: <int> | default = 1 ]
  # Maximum number of samples per send.
  [ max_samples_per_send: <int> | default = 100]
  # Maximum time a sample will wait in buffer.
  [ batch_send_deadline: <duration> | default = 5s ]
  # Initial retry delay. Gets doubled for every retry.
  [ min_backoff: <duration> | default = 30ms ]
  # Maximum retry delay.
  [ max_backoff: <duration> | default = 100ms ]
```
Based on an issue I found (https://github.com/prometheus/prometheus/issues/5166) I will attempt to adjust `max_samples_per_send` and see what happens.

**Comment by jakubgs** [comment]
I bumped these values in my settings:
```yaml
     queue_config: 
       capacity: 25000 
       max_samples_per_send: 5000 
```
And I now see more stable behavior and less `push` requests:
![cortex_requests_lower](https://user-images.githubusercontent.com/2212681/99684548-dfdecb00-2a81-11eb-9d51-f307d8b5664e.png)
Though I'm not sure exactly what I should be looking at to tune those further.

**Comment by jakubgs** [comment]
And yet, we're back to the same issue:

![cortex_same_issues_again](https://user-images.githubusercontent.com/2212681/99689469-53370b80-2a87-11eb-8369-29103af1d177.png)

Prometheus is not having it either:

![prometheus_retired_samples_spike](https://user-images.githubusercontent.com/2212681/99689397-3c90b480-2a87-11eb-851a-5f5df0951347.png)

**Comment by jakubgs** [comment]
The `cortex_ingester_flush_queue_length` metric going up must mean something, but not sure what.

And why is every time a single host in the cluster that has issues? Why none of the others?

**Comment by jakubgs** [comment]
And it's using all 6 CPUs, I don't know why:

![cortex_master_03_full_cpu](https://user-images.githubusercontent.com/2212681/99694467-c42cf200-2a8c-11eb-8920-9f4fc4e1ea02.png)

**Comment by jakubgs** [comment]
Why does the queue suddenly shoot up in [this image](https://user-images.githubusercontent.com/2212681/99689469-53370b80-2a87-11eb-8369-29103af1d177.png)? If it was a tuning issue shouldn't gradually grow and become a bigger and bigger problem? Rather than suddenly spike from 500 to 300K in 15 seconds. What is that about?

**Comment by fggarcia** [comment]
we have similar scenario

![image](https://user-images.githubusercontent.com/1306596/99710762-ba50c200-2a7f-11eb-9c24-25e5bc3efacd.png)

**Comment by fggarcia** [comment]
update: rollback to prometheus 2.21.0 (2.22.2) resolve the issue

**Comment by jakubgs** [comment]
Thanks a bunch @fggarcia. I have also downgrade to `2.21.0` and reverted the `query_config` changes and it seems to be fine so far. Appreciate you commenting on this.

**Comment by jakubgs** [comment]
It's still happening, but at least now it's not a runaway process, and it just spikes from time to time.

![cortex_issues_are_back_milder](https://user-images.githubusercontent.com/2212681/99851198-0cc4d800-2b7f-11eb-9aaa-b02341da7973.png)

So it's not as catastrophic as it was before, but I'd still like to know why suddenly samples spike from 12K on each host to 51K on one host.

**Comment by jakubgs** [comment]
What I'm trying to understand is why suddenly on the Prometheus instances generates this spike in storage shards to the max of 1000 and desired shards?

![prometheus_spike_storage_shards](https://user-images.githubusercontent.com/2212681/99998210-9a423b00-2dbe-11eb-815f-ff31a05fe8e8.png)

Which correlates with a big spike in number of connections and volume of traffic on the Prometheus node:

![prometheus_network_traffic_conn_spike](https://user-images.githubusercontent.com/2212681/99998349-d7a6c880-2dbe-11eb-88db-da23c15964ff.png)

And on the Cortex side I see a slow dip in number of samples from that host, and then a spike:

![cortex_dip_in_samples_total](https://user-images.githubusercontent.com/2212681/99998474-0755d080-2dbf-11eb-8455-970f019e02eb.png)

**Comment by jakubgs** [comment]
Okay, I managed to calm it down with combination of downgrading Prometheus to `2.21.0` and some fine tuning in `query_config`:
```yaml
      min_shards: {{ cortex_services.json | length }}
      capacity: 1500
      max_samples_per_send: 300
      max_backoff: 2s
      batch_send_deadline: 4s
```
It's looking much better now:

![cortex_clean_graphs](https://user-images.githubusercontent.com/2212681/100224826-a1338f80-2f1d-11eb-968e-c82193dd6a9f.png)

Thanks a bunch to @fggarcia for the downgrade comment. That made a big difference.

**Comment by jakubgs** [comment]
I think I'd consider this resolved, at least for me.

**Comment by fggarcia** [comment]
Hi i know that the issue was closed. But the problem persist until with version ```cortex 1.6.0``` & ```prometheus 2.24.0```

**Comment by fggarcia** [comment]
Hi @jakubgs do you have prometheus at 2.21.0 version until now? Do you upgrade to new prometheus version? I'm still having issue with this.

**Comment by jakubgs** [comment]
Yes. I have upgraded to `2.25.0` and I'm still seeing these failures. Like yesterday:

![image](https://user-images.githubusercontent.com/2212681/116965777-e9b25e80-acae-11eb-8bb5-f33f8aef2669.png)

Out of nowhere Prometheus starts increasing storage shards and starts retrying sending samples.
On the Cortex side we can see increase in both 400s and 500s and spikes in flush queue length:

![image](https://user-images.githubusercontent.com/2212681/116965868-2ed69080-acaf-11eb-9680-ac2f254cc5d0.png)

I'm still confused as to what's causing this.

Just restarting the Prometheus instances fixes it, but you can never know when it comes back again.

**Comment by jakubgs** [comment]
I have abandoned Cassandra in favor of Block storage and it's performing very well.

Truth be told considering the amount of time and money I wasted trying to get Cassandra to work as a backend I would recommend putting in deprecation warnings in documentation or even gradually dropping support, because it's just absurd how poorly it performed. Maybe there's a way to fine-tune it to make it work, but based on my experience it's extremely difficult and just an absolute waste of time.

To anyone reading this: __DO NOT USE CASSANDRA AS A CORTEX BACKEND__
